### DISCUSSION TITLE
Best way to gather relationships between entities (NER and REL models)

### DISCUSSION THREAD

**Discussion by fernandonjardim** [explosion/spaCy#10808] [Help: Model Advice]
I am trying to create a model using SPACY [to output this JSON, given this text](https://docs.google.com/document/d/1hgLdaDDYnCMn7JuZ9CJap_fQBVfh-Dj19IlWyTnSV64/edit?usp=sharing).

To approach this problem I've tried to combine a NER + REL model, but this is not quite good and takes a LOT of time labelling the entities.

Would you suggest a best approach on how could I tackle this challenge using the least labelling possible?

**Comment by ljvmiranda921** [comment]
Hi @fernandonjardim ,

For posterity, I advise you to copy the sample JSON in your post. Be sure to enclose it with triple-backticks so that it's more readable. The GDocs file may get deleted, or its access may change, and future users may not be able to see what the post is about.

For the problem itself, you'd definitely want to label at least a hundred or so samples to get a decent model. However, it seems that some of the entities can be obtained by simple business rules. A combination of that + a model might actually help. You can use spaCy's [Matcher](https://spacy.io/api/matcher) for creating some rule sets. You might also want to explore techniques like [weak supervision](https://github.com/explosion/projects/tree/v3/tutorials/ner_tweets) to reduce your labelling costs.

**Comment by fernandonjardim** [comment]
Hi @ljvmiranda921 ![

Thanks very much for your fast answer and tips!

So instead of having two models, one for NER and another one for Relationship between NERs (as I am currently doing it, following this [article](https://towardsdatascience.com/how-to-train-a-joint-entities-and-relation-extraction-classifier-using-bert-transformer-with-spacy-49eb08d91b5c)), you would recommend to go with only NERs, doing some auto labelling and use SPACY's Matcher, is it right?

In most of my corpus, the entities /tokens can be really spread out, as follows:

[image_2022-05-17_143613693](https://user-images.githubusercontent.com/31636793/168824912-bb88f8a1-bb29-492b-bba2-6f24bdd9db02.png)

On the example above the label "included" (third in blue), got several children ("optionals", in black), and some are really far away, in some cases in other paragraphs. Would the matcher still  work in this case?

What I guess I am missing, is how to use the matcher on this long distance cases (picture), with the "Hello, World" example, it seems easier haha

Also some spans can get more than a label. For instance, when I am describing the itinerary on "Day 2",  the span "schooner tour" belong to the hash "day_2" but also to the "not_included" one. Can the matcher get those nuances too?

Super thanks for your availability!

**Reply by ljvmiranda921** [reply]
> So instead of having two models, one for NER and another one for Relationship between NERs (as I am currently doing it, following this [article](https://towardsdatascience.com/how-to-train-a-joint-entities-and-relation-extraction-classifier-using-bert-transformer-with-spacy-49eb08d91b5c)), you would recommend to go with only NERs, doing some auto labelling and use SPACY's Matcher, is it right?

I'm thinking about using a combination of business rules and the NER for extracting the entities. Then once you have those entities, then you can apply whatever downstream tasks you need (getting relationships, etc.). 

> Also some spans can get more than a label. For instance, when I am describing the itinerary on "Day 2", the span "schooner tour" belong to the hash "day_2" but also to the "not_included" one. Can the matcher get those nuances too?

It might be possible to configure a very custom Matcher for that (i.e., create multiple rules that obtains the same thing). If you're performing entity extraction on overlapping entities, perhaps you can check out the [`spancat` component](https://spacy.io/api/spancategorizer).

**Comment by fernandonjardim** [comment]
Hi @ljvmiranda921 !

Thanks very much for your answer! 

I still a bit blocked on getting the relations, and when is better to use business rules and when is better to use NER.

I've been through all the spacy docs, and all of them use short sentences as an example. When you get a sentence such as the following, it is easy to create business rules using a matcher or do dependencies analysis:

"Apple is opening its first big office in San Francisco"

What I am not getting it is how can I use these tools in long sentences, in which the children / dependencies are spread around many sentences as the following example, copying the image as a text, this is a bit of our corpora:

```
REVEILLON 2022 ARRAIAL DO CABO DATE 27/12/2021 to 02/01/2022. PACKAGES INCLUDE Transport by executive bus All our partners have the certifications required by ARTESP and ANTT as well as day maintenance and passenger insurance. 6 nights accommodation @ a Hostel, during the whole period of the trip Accommodation at a Hostel with excellent location 3 minutes from Praia dos Anjos 5 minutes from Praia Grande 5 minutes from the trail to Praia do Forno. Near by restaurants bars markets pharmacies and other establishments. Our check-in will be immediate when we arrive at the place on the morning of the 28th. We do not need to follow the hotel system where the check-in is from 14.00. - breakfast included on 28 29 30 31 01 and 02. Welcoming to Arraial do Cabo Barbecue  On the first night in Arraial do Cabo we will have the welcome barbecue with a variety of salted meats and garnishes. NYE Party with Open bar Vodka gin catuaba beer energy drink refrigerant and water. Guided walks to Praia do Forno.
``` 

to extract this info:

`{"included": ["Transport by executive bus", "passenger insurance",  "6 nights accommodation @ a Hostel", "breakfast included on 28 29 30 31 01 and 02", "Welcoming to Arraial do Cabo Barbecue", "On the first night in Arraial do Cabo we will have the welcome barbecue with a variety of salted meats and garnishes.", "NYE Party with Open bar Vodka gin catuaba beer energy drink refrigerant and water", "Guided walks to Praia do Forno"]}
`

For this case, would you suggest NER labelling? Or some how still using business rules? If business rules, would you mind in providing just a very simple example, so I can replicate to all the other entities / relations I need to extract?

Super thanks.

**Reply by ljvmiranda921** [reply]
Hi @fernandonjardim ,

My advice would be to try both: (1) see if the business rules get atleast 70-80% of your use-cases then see if (2) NER can do the same. By looking at your text, it's still a bit unclear (or there's no consistent definition) what is considered as an entity, so you might get more mileage just trying out both and see what sticks. 

For example, the `"included"` tag doesn't seem to have a consistent pattern to train a model upon. As a naive example of a pure business rule approach, what might work is to look for texts after a specific keyword (or between two keywords) and then prune the output to get the thing that you want.

**Comment by fernandonjardim** [comment]
From a scale of 0 o 10, could you please roughly tell me how hard is this challenge?

I've been months on it and I just cannot ship it out in a simple way which is not NER + Relationship. 

Having things spread out the text is what is really making this hard. For instance sometimes I have things on the same chunk: "Boarding @ 11 @ Metro Vergueiro" and sometimes I have things on the following "bla bla bla bla bla bla boarding:  we will meet in front of the shopping and we will meet @ 11 @ metro Vergueiro" and this makes hard to come up with a scheme for it

**Reply by ljvmiranda921** [reply]
Hi @fernandonjardim, for the Discussions forum, instead of creating a new "Answer," you can "Reply" on the same thread 🙂 This makes reading the conversation more manageable (especially when an Answer has already been marked).

It's not straightforward to give a definite difficulty score to your problem as it depends on many factors. Although I'd say that writing a few business rules, although hard to generalize, might still help you, especially for the more "obvious" entities. So perhaps you're on the right track of thinking this problem as NER and REL, but you might want to consider adding a few business rules before or after that process.

**Reply by fernandonjardim** [reply]
Super thanks for the feedback and learnt about replying on the thread (:

**Reply by fernandonjardim** [reply]
Hi @ljvmiranda921 !

I've written some business rules, and I've been able to output the spans, but I cannot find ways to add it to the pipe, meaning, how to annotate data using them. 

For instance, as you know I already have a model working that predicts a few labels. How can I updated it to use the Matcher patterns?

Super thanks in advance

**Reply by fernandonjardim** [reply]
ok got it!
if I want to use matcher, I need to update the docs.ents manually, meanwhile the entity ruler does it automatically, right?
I was struggling with the Matcher and the docs.ents ... Think I sorted ((;

**Reply by polm** [reply]
The EntityRuler wraps the Matcher so you don't have to update `doc.ents` manually. If you want to combine your ruler with an NER pipeline, you can add it to an existing pipeline using `nlp.add_pipe`, for example. You may want to look at [this sample project](https://github.com/explosion/projects/tree/v3/tutorials/ner_double) for an explanation of how multiple pipes interact.

**Reply by fernandonjardim** [reply]
Super thanks @polm !

Quick Questions:

1) how do I see what is the order that things are being run into the pipe (and if I want to change the order between things into the pipe, what would be the best practice?)

2) I want to create a pipe with empty ents, so when I run the entity ruler, I can clearly see where they are being applied. What would you recommend?

Super thanks in advance!

**Reply by polm** [reply]
You can use `nlp.pipe_names` to see pipes in the order they will be run in. It may help to read the [pipelines section of the docs](https://spacy.io/usage/processing-pipelines) if you haven't seen it before.

I don't understand your second question. If you want to see what a pipe is doing, keep in mind a pipe is like a function that takes a doc and returns a doc, so you can also do something like `ruler(doc)` to see what changes the pipe would make.

**Reply by fernandonjardim** [reply]
Hi @polm !

Thanks very much for the reference, I've seen the pipelines section but It still a bit unclear for me.

For instance what I am trying to do, is to have a doc that only got the labels from the entity ruler I've created (in this case, the label is "Dates") without the other labels such as LOC; ORG; etc.

So I've tried to load a blank model and also loading a language model but disabling (also tried excluding) the NER module, so when I run my entity ruler and do the nlp(doc) it could apply only the ents from the ruler, but once I do nlp(doc)  it also  recognizes and applies the ents from the model (even tho it was a blank model or had the NER model disabled or Excluded) /:

**Reply by fernandonjardim** [reply]
Hi @polm, there was probably something going on with my runtime. 
I haven't touched the code since last night, just open up and now it works --' 
Thanks for your help (: 👍

**Reply by fernandonjardim** [reply]
Hi @polm !

I am trying to create a simple pattern to capture Brazilian money value. The very simple stuff, like R$ is not working /:

[Here is the link of an example](https://explosion.ai/demos/matcher?text=R%24%201.728%2053&model=en_core_web_sm&pattern=%5B%7B"id"%3A1%2C"attrs"%3A%5B%7B"name"%3A"LOWER"%2C"value"%3A"R"%7D%2C%7B"name"%3A"ORTH"%2C"value"%3A"%24"%7D%5D%7D%5D)

Do you know what could be wrong?

I also got cases like: R$ 10 or R$ 100 or R$ 1000 or R$ 1,532 23 .

For all these cenarios, do I need to create a specfic rule according to the number of characters?

Thanks in advance (:

**Reply by polm** [reply]
When you ask a question and say code doesn't work, please include a sample of code we can run to check for you.

This is the pattern from the Matcher demo you linked to:

```
pattern = [{'LOWER': 'R', 'ORTH': '$'}]
```

In a Matcher pattern, each dictionary is one token. So this describes one token where the **whole** text (ORTH) is "$" and the same text in lowercase is "R". It is not possible for such a token to exist (and "R" is not lowercase). 

It looks like "R$" is one token, so `"ORTH": "R$"` should work.

**Reply by fernandonjardim** [reply]
Hi @polm , thanks very much for your reply! 

Sorry for not pasting the code, I thought the pattern was going with the link I've shared. I was trying to capture Captalized R or Lower R, I've added the `"Lower": "R",  "OP: !" ` and it worked (:

Not what is the best way to match values? I can get R$ 10, or R$ 135 or R$ 12.000 .

Should I create 5 or 6 "shapes", each one for the number of characters? 

Super thanks for all the support (: 

Last but not least, Instead of doing NER + Relationship models  I was wondering about using SPANCAT, as I have a some labels that they overlap.

Having in consideration all the discussion above with @ljvmiranda921 , do you think this may be a better strategy than NER + Relationship models?

**Reply by fernandonjardim** [reply]
Hi there, 

Adicionally, this is one of my patterns (trying to only capture `R$750` or `r$ 750` :

```pattern_money` = [{"label": "MONEY", "pattern": [{'LOWER': 'R', "OP": "!", 'ORTH': '$'}, {'IS_DIGIT': True}], "id": "3"},
                 {"label": "MONEY", "pattern": [{'IS_DIGIT': True}, {'IS_DIGIT': True, "OP": "?"}, {'ORTH': 'reais'}], "id": "4"}]```


and I am running it over this text:

`Foz do Iguaçu Paraguai Data : 20 á 25/04/22 Feriado A partir de R$ 750 00 por pessoa.Se tivermos a Liberação das Fronteiras até a data Incluiremos a Argentina no nosso Roteiro. INFORMAÇÕES DE EMBARQUE E RETORNOTATUAPÉ: APRESENTAÇÃO : 19:30 SAÍDA : 19:45 TOLERÂNCIA MÁXIMA PONTO DE ENCONTRO : Rua Domingo Agostin 134 MONEY Atrás do Shopping Metrô Tatuapé.BARRA FUNDA : APRESENTAÇÃO : 20:00 SAÍDA : 20:15 TOLERÂNCIA MÁXIMA PONTO DE ENCONTRO : Rua Tagipuru 644 Portão 02 do Memorial.Retorno: 24/04/2022 `

but it tags :  `20`,  `Agostin 134`, `Tagipuru 644`,  `Portão 02` and I don't understand why /:

**Reply by fernandonjardim** [reply]
Also can't understand why this pattern: `[{'ORTH': 'R$'}, {'IS_DIGIT': True}, {'IS_DIGIT': True}]` doesn't capture `R$1800 00.`, once if I run `print([token.text for token in tokens])` I get 4 different tokens '['R$', '1800', '00', '.']'

**Reply by polm** [reply]
Hey, like I mentioned above, please include your code. Without your full code it's not clear what's going on. Please share a full piece of code we can copy/paste and run.

Also your use of `"OP": "!"` is wrong. Forgive me for not being clearer in my explanation. For this rule:

```
[{'LOWER': 'R', 'ORTH': '$'}]
```

This will match text where `text.lower() == "R" and text == "$"`. That is not possible because "R" is not lowercase and a string cannot be "R" and "$" at the same time. I think you figured that out but I want to be clear. The "!" match means something like `not(other conditions)`, so if you have `not(impossible conditions)` that will match any token.

This code produces a match for me.

```
import spacy
from spacy.matcher import Matcher

nlp = spacy.blank("pt")
pattern = [{"LOWER": "r$"}, {"IS_DIGIT": True}, {"IS_DIGIT": True}]
matcher = Matcher(nlp.vocab)

matcher.add("MONEY", [pattern])

doc = nlp("R$1800 00.")

print("TOKENS")
for tok in doc:
    print(tok)
print("MATCHES")
for _, start, end in matcher(doc):
    print(doc[start:end])
```

**Reply by fernandonjardim** [reply]
Super thanks @polm !

I still not getting how the thing works.

For instance here: `R$ 1.728 53`  `550$` :


```
import spacy
from spacy.matcher import Matcher

nlp = spacy.blank("pt")
tokens = nlp("R$ 1.728 53")
tokens2 = nlp("550$")

````

Here I will get 3 tokens: 

`['R$', '1.728', '53']`

Here I get 1 token:

`['550$']`



so:


```
ruler = nlp_test.add_pipe("entity_ruler")
new_rule = [{"label": "MONEY", pattern = [{'LOWER': 'r$'}, {'IS_DIGIT': True, 'IS_PUNCT': True, 'SHAPE': 'dddd'}, {'SHAPE': 'dd'},
{"label": "MONEY", pattern = [{'IS_DIGIT': True, 'ORTH': '$'},
{"label": "MONEY", pattern = [{'IS_DIGIT': True}, {'ORTH': '$'}]}]
ruler.add_patterns(new_rule)
tokens = nlp("R$ 1.728 53  550$")
tokens .ents
```


And I don't get the `R$ 1.728 53` tagged. Do you know why? With my rule I am trying to capture first token (R$), second token which is a number followed by a point, and 4 digits, and third token which got 2 digits.

With `550$`  I've tried to capture digits than $, single token or multiple token.

Also, my pattern for money varies o lot. I can have tokens with 1, 2, 3 or 4 shapes. Is there anything I can do like "up to" 4 digits for instance?

Super thanks in advance!

**Reply by polm** [reply]
I'm not sure about this, but you seem to be confused about how the attributes in a dictionary work. Each of the attributes describes **the whole token**. So when you say `"ORTH": "$"`, that is like `text == "$"`, not like `"$" in text`. So it won't match a token like `R$` or `500$`. When you say `"IS_DIGIT": True`, that means **every** character is a digit, not just some of them. This code snippet may be helpful:

```
import spacy
nlp = spacy.blank("pt")
doc = nlp("R$ 1.728 53  550$")

for tok in doc:
    print(tok, tok.is_digit, tok.is_punct, sep="\t")
```

A couple of things to note here:

1. `is_digit` is True for `53`, but not for `1.728` or `550$`
2. When you have two spaces in a row, one of them becomes a token
3. `is_punct` is False for all tokens

If you need to match tokens like `1.728` I would use a regex pattern, like `"REGEX": "^[0-9.]+$"`. For example this works:

```
import spacy
from spacy.matcher import Matcher

nlp = spacy.blank("pt")
pattern = [{"LOWER": "r$"}, {"TEXT": {"REGEX": "^[0-9.]+$"}}, {"IS_DIGIT": True, "OP": "+"}]
matcher = Matcher(nlp.vocab)

matcher.add("MONEY", [pattern])

doc = nlp("R$ 1.728 53  550$")

print("TOKENS")
for tok in doc:
    print(tok, tok.is_digit, tok.is_punct, sep="\t")
print("MATCHES")
for _, start, end in matcher(doc):
    print(doc[start:end])
```

> Also, my pattern for money varies o lot. I can have tokens with 1, 2, 3 or 4 shapes. Is there anything I can do like "up to" 4 digits for instance?

There is a PR for that (#10981), but no such feature at the moment. You can generate patterns of all relevant lengths, or you can use one of the operators like `+` or `*` to match an unlimited number of tokens, which is probably fine for currency patterns.

**Reply by fernandonjardim** [reply]
Hi @polm !

Thanks very much for the detailed explanation!

I still with some issues here

For instance, I am trying to pick time patterns, [as you can see here](https://explosion.ai/demos/matcher?text=22h22&model=en_core_web_sm&pattern=%5B%7B%22id%22%3A0%2C%22attrs%22%3A%5B%7B%22name%22%3A%22SHAPE%22%2C%22value%22%3A%22ddhdd%22%7D%5D%7D%5D) for '22h22', i can create a `pattern = [{'SHAPE': 'ddhdd'}]`  and it works.

But I am trying the following code and it doesn't recognize it:

 
```
nlp_test = spacy.blank("pt")
pattern_money = [{"label": "TIME", "pattern": [{'SHAPE': 'ddhdd'}], "id": "1"}]
ruler_test = nlp_test.add_pipe("entity_ruler")
ruler_test.add_patterns(pattern_money)
ruler_test.patterns
-> [{'id': '1', 'label': 'TIME', 'pattern': [{'SHAPE': 'ddhdd'}]}]
tokens = nlp_test("22h22")
print([token.text for token in tokens])
-> 22h22
displacy.render(tokens, style="ent", jupyter=True)
-> 22h22
```
 
It does print the token, but not with the label /:

**Reply by polm** [reply]
When you paste code that you want us to debug, please paste a whole sample we can copy and run ourselves, with imports and everything. That makes it easier for us to check your code.

When I ran your code with imports, I got this warning:

> /mnt/pool/work/spacy/env/lib/python3.10/site-packages/spacy/displacy/__init__.py:205: UserWarning: [W006] No entities to visualize found in Doc object. If this is surprising to you, make sure the Doc was processed using a model that supports named entity recognition, and check the `doc.ents` property manually if necessary.
  warnings.warn(Warnings.W006)

This means there are no entities in the Doc. Did you not see that warning? I'm not sure what could cause it to not be shown...

The issue here is your shape is wrong. Since you're trying to match on `tok.shape`, you can debug that by printing the shape. 

```
print(tokens[0].shape_)
# => ddxdd
```

Here you see that the "h" has become an "x". If I update the pattern to reflect that it just works.

**Reply by fernandonjardim** [reply]
Hi @polm  thanks very much for your feedback!
Yes, I do get the error message and that is what is confusing me, once if I go to the online Matcher's demo, with the shape ' ddhdd' it matches the token...
So everytime I have any letter on my shape, I should use x right?

**Reply by polm** [reply]
Not sure what's going on with the web demo, but you shouldn't use an `h` there. `x` is for lowercase letters, `X` for upper. If in doubt check the actual shape of the token you're trying to match. Note that word shape is mainly a feature for the tok2vec and not a tool for matching; while you can use it that way, regexes or other tools will give you more options.

**Reply by polm** [reply]
Note: I figured out what was up with the web demo - it was highlight all entities, not just entities created by matching the rules. So the example you had was being matched as a QUANTITY or something. This issue with the web demo has been resolved.

**Reply by fernandonjardim** [reply]
Super thanks @polm 

I am having some issues with PhraseMatcher, could you give me a hand?

I am trying to approach the same problem, but using the PhraseMatcher instead of the Entity Ruler.

My goals is to return a doc in which `22h22` is tagged. , but it seems that it has not been added to the pipeline.

Here is my full code 

```

#installs
!python -m spacy download pt

#imports
import spacy 
from spacy import displacy
from spacy.matcher import Matcher
from spacy.matcher import PhraseMatcher

#creating doc
nlp_test_time = spacy.blank("pt")
doc = nlp_test_time("the time now is 22h22")
doc.ents
# gets () here

#adding PhraseMatcher to pipe
matcher_time = PhraseMatcher(nlp_test_time.vocab, attr="SHAPE")
matcher_time.add("TIME", [nlp_test_time("22h12")])

#applying pipeline to doc
nlp_test_time(doc)
nlp_test_time.ents
# gets () here too, when I thought I was going to get `22h12`
```

Here I was expecting to get `22h12` as `TIME`

Do you know what I could be doing it wrongly?

Super thanks in advance (:

**Reply by polm** [reply]
```
#adding PhraseMatcher to pipe
matcher_time = PhraseMatcher(nlp_test_time.vocab, attr="SHAPE")
```

This doesn't add a PhraseMatcher to your pipeline, it just creates a PhraseMatcher. To add to your pipeline you need to call `nlp.add_pipe`, but the PhraseMatcher isn't a component. You can use the [EntityRuler](https://spacy.io/usage/rule-based-matching#entityruler), which wraps the PhraseMatcher.

**Reply by fernandonjardim** [reply]
Hi @polm 

Thanks very much, couldn' t get then how do I add the PhraseMatcher into the ruler still.. What I am trying to accomplish is to have a doc with the entities, not just the entities....

```
ruler = nlp_test_time.add_pipe("entity_ruler")
matcher_time = PhraseMatcher(nlp_test_time.vocab, attr="SHAPE")
matcher_time.add("TIME", [nlp_test_time("22h12")])
ruler.add_patterns("TIME")
```

Do I need a decorator or something like that ?

**Reply by polm** [reply]
You don't create the PhraseMatcher manually - the EntityRuler has a PhraseMatcher internally. Please read the EntityRuler documentation - you can set a [`phrase_matcher_attr`](https://spacy.io/api/entityruler#config), and then any strings you pass will match using the internal EntityRuler.

In this case, since you're only matching one token, you can just use a normal Matcher rule in the EntityRuler though.

```
import spacy

nlp = spacy.blank("en")

ruler = nlp.add_pipe("entity_ruler")
ruler.add_patterns([{"label": "TIME", "pattern": [{"SHAPE": "ddxdd"}]}])

doc = nlp("The time is 12h08 23")
for ent in nlp("The time is 12h08").ents:
    print(ent.label_, ent, sep="\t")
```

**Reply by fernandonjardim** [reply]
Super thanks for all your support @polm you rock!
if you ever in lisbon let me know, need to pay you some beers haha

**Comment by fernandonjardim** [comment]
Super thanks for the feedback polm ((:
